### DISCUSSION TITLE
When 'RuntimeError: The 'data' object was created by an older version of PyG.' evokes?

### DISCUSSION THREAD

**Discussion by songsong0425** [pyg-team/pytorch_geometric#4502] [Q&A] [answered]
Dear PyG team,
Greetings, I have a question about the RuntimeError.
While running the code from paper, I counted 'RuntimeError: The 'data' object was created by an older version of PyG' error.
I confirmed the link https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/data/data.py that looks related to this error, but I could not understand when this error evoke.
As described in the comment, I checked the dataset's root folder to remove the 'processed/' directory, but it didn't exist in my folder anywhere.
My PyG version is 2.0.3 and running with jupyter notebook, python 3.9.7, and pytorch 1.10.0.
Can anyone deal with this error?
Sincerely,
Songyeon

**Comment by rusty1s** [comment, marked as answer]
The `data` format changed slightly in `torch-geometric==2.0`. Thus, we cannot longer load in the data that were created by previous PyG versions. I am really sorry for the inconveniences! Once thing you can do is to re-process your to fit into the PyG 2.0 format. That is:
```python
# Install PyG 1.7.2
data = torch.load('data.pt').to_dict() 
torch.save(data, 'data_dict.pt')

# Install PyG 2.0
data_dict = torch.load('data_dict.pt')
data = Data.from_dict(data_dict)
torch.save(data, 'data.pt')
```

**Reply by songsong0425** [reply]
Dear rusty1s,
Thank you for your answer!
But does it affect to the code that open the .csv file for loading data?
Where should I modify the code as you suggested?
(Sorry for stupid question. Unfortunately, I'm not familiar with the code because it is not mine.)

**Reply by rusty1s** [reply]
This error only happens in case you load a pre-processed dataset or data object. It should not happen in case you re-process the data from scratch (e.g., in loading a `*.csv` file to create a new `data` object). The idea would be to apply this modification to the `data.pt` object inside `{dataset_dir}/processed`.

**Reply by songsong0425** [reply]
Thank you for your kind reply! Although I could not find the processed folder yet, I'll try other way. Thanks!

**Reply by eason1021** [reply]
Dear rusty1s,

I also met the same problem of my old dataset, and I found the post. But I also have some questions, in your example, what's the module of the 'torch' in PyG 1.7.2? Pytorch or Pytorch-Geometric? If the answer is pytorch, what's the function of PyG 1.7.2? Could I just use the PyG 2.0 to achieve the dataset transformation?

Thanks.

**Reply by rusty1s** [reply]
`torch` refers to PyTorch in the example. So what you can do is downgrade to PyG 1.7.2, do the transformation, and upgrade to PyG 2.* again.

**Reply by eason1021** [reply]
Dear rusty1s,

Thanks for your kind reply! 
However, when I downgrade to PyG 1.7.2, I cannot use the .to_dict(), it would show the Attribute error like: AttributeError: 'tuple' object has no attribute 'to_dict'. How could I resolve this issue?

**Reply by rusty1s** [reply]
Try
```python
data, slices = torch.load('data.pt')
data_dict = data.to_dict()

**Reply by eason1021** [reply]
Dear rusty1s,

It occurs new attribute error: AttributeError: 'Data' object has no attribute 'to_dict'. Did you mean: 'from_dict'?

I think this is weird, I also print the data, here is the information: 
Data(edge_index=[2, 184764384], test_mask=[26394912], train_mask=[26394912], val_mask=[26394912], x=[26394912, 21], y=[26394912])

**Reply by rusty1s** [reply]
`Data.to_dict()` should exist in 1.7.2, see https://github.com/pyg-team/pytorch_geometric/blob/1.7.2/torch_geometric/data/data.py#L113.

However, you can also do `data_dict = {key: data[key] for key in data}`.

**Reply by eason1021** [reply]
Dear rusty1s,

Thanks, the code execute successfully in PyG 1.7.2 with 
`data, slices = torch.load('data.pt')`
`data_dict = data.to_dict()`

However, when I do the following code in PyG 2.0:
`data_dict = torch.load('data_dict.pt')`
`data = Data.from_dict(data_dict)`
`torch.save(data, 'data.pt')`

The code can execute with no error message, however, I found I could not print the dataset information with new pt file and PyG Inmemory Dataset function:

If I print(len(dataset)), the error message would appear: AttributeError: 'tuple' object has no attribute 'items'

**Reply by rusty1s** [reply]
Wouldn't you want to save the tuple instead of the `data` object?
```
torch.save((data, slices), 'data.pt')

**Comment by tonyriverms** [comment]
A possible simpler solution from my experience might be directly reading from `__dict__`. All data seem actually saved there. In this way the Data object seems now compatible with the old version at least for reading in my use case.

For example,

```
from torch_geometric.data.data import Data
from typing import Mapping

def is_new_data_object(data: Data):
    return '_store' in data.__dict__

data: Data = pickle.loads(self.env.begin().get(f"{idx}".encode()))
if not is_new_data_object(data):
    data: Mapping = data.__dict__
return dict(
            pos=torch.as_tensor(data["pos"]).float(),
            pos_relaxed=torch.as_tensor(data["pos_relaxed"]).float(),
            cell=torch.as_tensor(data["cell"]).float().view(3, 3),
            atoms=torch.as_tensor(data["atomic_numbers"]).long(),
            tags=torch.as_tensor(data["tags"]).long(),
            relaxed_energy=data["y_relaxed"],  # python float
   )
```
